requested walk through of pacman_mirrors.py

fhdk opened the pull request:
done some refactoring

Several large function has been separated into logical smaller maintainable units

There is an error in pacman_mirrors_gui which prevents interactive

I have documented the code intensively since I had to get an understanding of the inner works before I could contribute to the code.

Huluti commented:
I like your changes. I've also made a PR to refactor the script yesterday. It could be a good idea to try to take the best ideas from the two and do only one PR (to avoid conflicts). What do you think?

fhdk commented:
lør, 07 01 2017 kl. 12:28 -0800, skrev Hugo Posnic:
> I like your changes. I've also made a PR to refactor the script
> yesterday. It could be a good idea to try to take the best ideas from
> the two and do only one PR (to avoid conflicts). What do you think?

Absolutely - will you do merge them?

I will have a look to see what I can learn from you :)

fhdk

Huluti commented:
I have received your email. What you can do is to commit here. We will work only in this PR.

fhdk commented:
Revised version committed - please review

Huluti commented:
Very great! Can you try to [make tests to be passed?](https://travis-ci.org/manjaro/pacman-mirrors/builds/190017773)

fhdk commented:
fixed tests to run local
removed utf-8 in all files
other minor improvements
i can get tests to run local but the travis check i dont know how to trouble shoot
the travis check complaints about missing `/etc/pacman-mirrors.conf` file and `/etc/pacman.d/mirrors` dir

test result
```
$ python setup.py test
running test
running egg_info
writing pacman_mirrors.egg-info/PKG-INFO
writing dependency_links to pacman_mirrors.egg-info/dependency_links.txt
writing top-level names to pacman_mirrors.egg-info/top_level.txt
reading manifest file 'pacman_mirrors.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '__pycache__' found under directory '*'
writing manifest file 'pacman_mirrors.egg-info/SOURCES.txt'
running build_ext
test_geoip_country_available (tests.test_pacman_mirrors.TestPacmanMirrors) ... :: Randomizing server list...
ok
test_geoip_country_not_available (tests.test_pacman_mirrors.TestPacmanMirrors) ... :: Randomizing server list...
ok
test_run (tests.test_pacman_mirrors.TestPacmanMirrors) ... :: Randomizing server list...
ok
test_run_country (tests.test_pacman_mirrors.TestPacmanMirrors) ... :: Randomizing server list...
ok

----------------------------------------------------------------------
Ran 4 tests in 0.006s
```

fhdk commented:
apparently the travis build did not recognize self.config["mirror_lists_dir"]
fixed - all checks passed

Huluti commented:
Good work! Version has a little problem in __init__.py and please use #!/usr/bin/env python and not #!/usr/bin/env python3 because python is by default python3.

Huluti commented:
I've an error when running: sudo ./pacman_mirrors.py -g

```
:: Interrogation des serveurs, cela peut prendre quelques temps...
=> Custom
Error: Cannot read file '/etc/pacman.d/mirrors/Custom': No such file or directory

Erreur: Aucun serveur disponible !
```

It tries to read /etc/pacman.d/mirrors/Custom instead of /var/lib/pacman-mirrors/Custom.

fhdk commented:
I don't. Have you checked your `pacman-mirrors.conf`

I have copied the files from the working project 
`pacman_mirrors.py` and `pacman_mirrors_gui.py`
to
`/usr/lib/python3.6/site-packages/pacman_mirror/`
did a ls to ensure no files exists at Custom
and did a `pacman-mirrors -g`
this is the output
```
$ ls -la /var/lib/pacman-mirrors
ls: cannot access '/var/lib/pacman-mirrors': No such file or folder
$ls -la /etc/pacman.d/mirrors/Custom
ls: cannot access '/etc/pacman.d/mirrors/Custom': No such file or folder
$ sudo pacman-mirrors -g
:: Querying servers, this may take some time...
=> Africa
=> Australia
=> Austria
=> Belarus
=> Belgium
=> Brasil
=> Bulgaria
=> Canada
=> Chile
=> China
=> Colombia
=> Costa_Rica
=> Czech
=> Denmark
=> Ecuador
=> France
=> Germany
=> Greece
=> Indonesia
=> Ireland
=> Italy
=> Japan
=> Netherlands
=> Philippines
=> Poland
=> Portugal
=> Romania
=> Russia

Error: Failed to reach the server: ftp error: timeout('timed out',)
=> Singapore
=> Sweden
=> Taiwan
=> Turkey
=> United_Kingdom
=> United_States
=> Vietnam
:: Writing mirrorlist ...
:: Generated and saved '/etc/pacman.d/mirrorlist' mirrorlist.
$ 

```

Huluti commented:
[Here's my conf.](https://framabin.org/?54d21e4ac4eff662#IsJsA4nocyW7hD2p+Dod/wUY6ASrYiYEpnWRJR2eZ0I=)

fhdk commented:
Your conf is ok - a case error has slipped in - see the commit.

This has revealed another problem

It seems like this line is missing the mirror-file name

self.config["only_country"] = [os.path.basename(self.custom_path)]

Huluti commented:
Do what you think it's better. You know more the code than me now ;)

fhdk commented:
this line

`if os.path.isfile(self.custom_path):`
afais this is a folder - which would explain why it drops a warning and query all servers

I dont know if this is the right way code-wise or it is just a cheap work-around
but if i change the code to this - it works as expected
```
if os.path.isfile(self.custom_path + self_config["only_country"]):
                    # it does
                    self.config["only_country"] = [os.path.basename(self.custom_path) + self.config["only_country"]]
                    self.debug_print("load_server_lists", "config['only_country']", self.config["only_country"])
                    # use it
                    self.mirror_dir = os.path.dirname(self.custom_path)
```
I will commit as is
Please review my code - did I mention - never touched python before?
And change obvious newbie constructions to best practice

fhdk commented:
not working got another err

Huluti commented:
Doesn't think that is working:
self.config["only_country"] = [os.path.basename(self.custom_path) + self.config["only_country"]]
os.path.basename(self.custom_path) is not needed anymore. It returned Custom when self.custom_path as the file and not a dir.

fhdk commented:
so self.custom_path is the complete path - so `/Custom` **is** the mirror file

fhdk commented:
@Huluti 
I have now tweaked and polished and hunted down all the little bugs I could.

Could you please review the latest changes

* verify the contents of `/etc/pacman-mirrors.conf`
* verify the contents of `/etc/pacman.d/mirrorlist` and `/var/lib/pacman-mirrors/Custom`
-- when using `-g --country country,country` 
-- when using `-g --interactive`
-- when using `-g`
-- when using `-g -c all`
* try --verbose
* try --notred

Huluti commented:
Very great, all is working for me now. It's --nored and not --notred ;). Let me a little time to make a review of the code.

Huluti commented:
Just check my conf file, and there is something strange, I have: OnlyCountry = '/var/lib/pacman-mirrors/Custom'. Shouldn't be just 'Custom'?

fhdk commented:
tir, 10 01 2017 kl. 07:06 -0800, skrev Hugo Posnic:
> Just check my conf file, and there is something strange, I have:
> OnlyCountry = '/var/lib/pacman-mirrors/Custom'. Shouldn't be just
> 'Custom'?

I have been very thorough but something obviously slipped.

How did you get there?

Huluti commented:
It comes from here:
```
if self.config["only_country"] == ["Custom"]:
    # check if such a 'Custom' file exist
    if os.path.isfile(self.custom_mirror_file):
        # it does
        self.config["only_country"] = [self.custom_mirror_file]
```

fhdk commented:
That's correct.

But in `generate_mirror_list_interactive()`

the value is restored to "Custom"

    self.config["only_country"] = ["Custom"]

Huluti commented:
So is this normal that I've this in my conf file `/var/lib/pacman-mirrors/Custom`?

fhdk commented:
No - its not normal - Can you try to reproduce the steps - because no matter what I do - I have "Custom".

Please try a `pacman-mirrors -g -c all`

and then try create a custom file like you did when the result was the complete file in config

Huluti commented:
It's working now, maybe an old conf from a commit with a bug that is already fixed.

fhdk commented:
Indeed there was a bug which did excactly what you describe and it has been squashed

fhdk commented:
the function `config_set_custom()` has a piece which ensures the correct output to config

```
for country in config_only_country:
    if "Custom" in country:
        country = "Custom"

   new_config = ("OnlyCountry = {country_list}\n").format(
        country_list=",".join(config_only_country))
```

Huluti commented:
Great! I think ll is ready. What do you think @Kirek?

fhdk commented:
I have looked over the code and something did spring to mind

[Issue #1](https://github.com/manjaro/pacman-mirrors/issues/1)
[Issue #10 ](https://github.com/manjaro/pacman-mirrors/issues/10)

Seems to have solved themselves with this code.
Otherwise I am very satisfied with the result - especially because I have no previous experience with python.

**Thank you for letting me have the opportunity to learn from you.**

Kirek commented:
Seems mostly fine, I'm going to fully review this now.

Kirek commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -127,211 +147,361 @@ def parse_cmd(self):
                             action="store_true",
                             help=_("detect country using geolocation, "
-                                   "ignored if '--country' is used"))
+                                    "ignored if \"--country\" is used"))
```
revert to single quotes

fhdk replied:
try comparing 
`ignored if '--coutnry' is used`
with
`ignored if "--country" is used`

Grammatically double qoutes is used when you qoute a message relayed by a person

He said "Bring me that car" in opposition of qouting a wording like the above text

Kirek replied:
you're right

Kirek commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -127,211 +147,361 @@ def parse_cmd(self):
-                            help=_("don't generate mirrorlist if NoUpdate is "
-                                   "set to True in the configuration file"))
-        if gtk_available:
+                            help=_("don\"t generate mirrorlist if NoUpdate is "
```
single quotes

fhdk replied:
this single qoute is not replaceable `don"t` does not work in English

Kirek replied:
the new code changed it to don"t

Kirek commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -447,175 +608,368 @@ def random_servers(self, countries):
+            elif hasattr(err, "code"):
+                # log error number to console
+                if verbose:
+                    print("\n" + _("Error: The server couldn\"t "
```
single quotes

fhdk replied:
The same
`Error: The server couldn"t ` is not correct grammar
Correct:
`Error: The server couldn't `

Kirek commented:
There is a bug when using interactive and random method, the gui list is empty.

I see that you slightly changed the behaviour when using the --country, now we always save the country. I'm okay with this.

There are too many comments in the code, sometimes commenting things that are obvious, like commenting we are going to print an error just before a print(...) function. Comments are fine but try to remove ones like that.

I don't like --nored switch, we already kind of doing this automatically by adding servers with longer than 24h since last sync to the bad server list. In fact I don't like the whole colour thing in pacman-mirrors, even if repo.manjaro.org does that. Keeping with the 3 categories - good, bad(not sync in +24h) and no response is enough. 
The reason is that this may exclude perfectly good mirrors with good ping to some users because they only sync 1-3 times a day. And they are good enough for our users in the default stable branch because they receive updates every week or two.

I see that you changed when writing the mirror-list from "## Location" to "## Country", this is fine as it's more consistent with everything else. But you need to take into account that some files will still have "Location" written in them so you need to modify get_mirror_country() function to take that into account.

I think that's all I could find :)

Kirek commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -447,175 +608,368 @@ def random_servers(self, countries):
+        elif line.startswith("## Country"):
+            # get it
+            (key, value) = line.split(":", 1)
+            key = key.rstrip()
```
No need to strip the key if you won't use the variable

fhdk replied:
That's right - it was a cpy/pst - will correct

fhdk commented:
tir, 10 01 2017 kl. 21:01 -0800, skrev Ramon Buldó:
> There is a bug when using interactive and random method, the gui list
> is empty.

I will have a look at that

> I see that you slightly changed the behaviour when using the --
> country, now we always save the country. I'm okay with this.
> There are too many comments in the code, sometimes commenting things
> that are obvious, like commenting we are going to print an error just
> before a print(...) function. Comments are fine but try to remove
> ones like that.

You're right

> I don't like --nored switch, we already kind of doing this
> automatically by adding servers with longer than 24h since last sync
> to the bad server list. In fact I don't like the whole colour thing
> in pacman-mirrors, even if repo.manjaro.org does that. Keeping with
> the 3 categories - good, bad(not sync in +24h) and no response is
> enough.
> The reason is that this may exclude perfectly good mirrors with good
> ping to some users because they only sync 1-3 times a day. And they
> are good enough for our users in the default stable branch because
> they receive updates every week or two.

That's fair and your logic is non-arguable :)

> I see that you changed when writing the mirror-list from "##
> Location" to "## Country", this is fine as it's more consistent with
> everything else. But you need to take into account that some files
> will still have "Location" written in them so you need to modify
> get_mirror_country() function to take that into account.
> I think that's all I could find :)

I will change the code accordingly

Thanks for the review

fhdk commented:
There is something I would ask you to check.

```
pacman-mirrors -i -m random # just click 'show list' and 'done
:: Querying servers, this may take some time...
=> Mirrors in Africa
=> Mirrors in Australia
<snipped>
=> Mirrors in Taiwan
=> Mirrors in Turkey
=> Mirrors in United_Kingdom
=> Mirrors in United_States
=> Mirrors in Vietnam

:: User generated mirror list
--------------------------
:: Writing custom mirror file ...
:: Generated and saved '/etc/pacman.d/mirrorlist' mirrorlist.
:: Writing mirror file ...
:: Saved personalized list of mirrors in '/var/lib/pacman-mirrors/Custom'.
```

output is shortend for brevity
```
pacman-mirrors -g -b unstable --verbose
:: Querying servers, this may take some time...
=> Custom mirror list

Error: Failed to reach the server: Not Found
==> Denmark -  0.005 - http://www.uex.dk/repos/manjaro/stable/$repo/$arch

Error: Failed to reach the server: Not Found
==> Netherlands -  0.031 - http://ftp.nluug.nl/pub/os/Linux/distr/manjaro/stable/$repo/$arch

Error: Failed to reach the server: Not Found
==> Netherlands -  0.039 - http://ftp.snt.utwente.nl/pub/linux/manjaro/stable/$repo/$arch

Error: Failed to reach the server: timed out
==> Chile -  2.003 - http://doge.ing.puc.cl/Mirrors/Manjaro/stable/$repo/$arch

Error: Failed to reach the server: ftp error: timeout('timed out',)
==> Russia -  2.053 - ftp://kibo.remi.lu/pub/stable/$repo/$arch
:: Writing mirror list ...
==> Denmark : http://www.uex.dk/repos/manjaro/stable/$repo/$arch
==> Netherlands : http://ftp.nluug.nl/pub/os/Linux/distr/manjaro/stable/$repo/$arch
==> Netherlands : http://ftp.snt.utwente.nl/pub/linux/manjaro/stable/$repo/$arch
==> France : http://manjarolinux.polymorf.fr/stable/$repo/$arch
==> Russia : ftp://kibo.remi.lu/pub/stable/$repo/$arch
:: Generated and saved '/etc/pacman.d/mirrorlist' mirrorlist.
```

The custom mirrorfile is generated for stable repo.
But when generating the mirrorlist I asks for unstable which gives error.

I think the custommirror file should have `$branch` instead of the actual branch.
Am I correct in assuming I have made an error here since a named branch is output to the mirror file?

Kirek commented:
Yes, `/var/lib/pacman-mirrors/Custom` should have $branch instead of the actual branch, like the other country files. I can confirm that this bug was introduced in this PR.
/etc/pacman.d/mirrorlist is written correctly

fhdk commented:
ons, 11 01 2017 kl. 03:15 -0800, skrev Ramon Buldó:
> Yes, /var/lib/pacman-mirrors/Custom should have $branch instead of
> the actual branch, like the other country files. I can confirm that
> this bug was introduced in this PR.
> /etc/pacman.d/mirrorlist is written correctly

Good - I have worked it through an have a commit ready.

fhdk commented:
I will remove the response time and the last sync time from the custom mirror file since the info is useless.

```
## Country           : Germany
## Response Time     : 99.99
## Last sync (hh:mm) : 99:99
Server = http://ftp.halifax.rwth-aachen.de/manjaro/$branch/$repo/$arch
```

Huluti commented:
@fhdk I'm sorry but I don't know if there is a reason for this.

Huluti commented:
Oops, you've integrated the UI PR in your last commit..

fhdk commented:
Ups - how do I revert it?

Huluti commented:
Just made make a new commit. I do not think that Kirek will merge 60 commits. He will make a squash for the creation of a single commit.

Huluti commented:
If we try to follow the [versioning conventions](http://semver.org/), I think that changes that has been realized require at least a 2.1 version or make a bump to 3.0 and not 2.0.1.

Huluti commented:
Why not adding also a zfill(2) to last_sync? It makes sense because http://repo.manjaro.org/ use also this format.
`datesync = '{}:{}'.format(str(int(hours)).zfill(2), str(int(minutes)).zfill(2))`
Moreover, datesync is not a good var name because it's not a date.

fhdk commented:
ons, 11 01 2017 kl. 13:17 -0800, skrev Hugo Posnic:
> Why not adding also a zfill(2) to last_sync? It makes sense because h
> ttp://repo.manjaro.org/ use also this format.
> datesync = '{}:{}'.format(str(int(hours)).zfill(2),
> str(int(minutes)).zfill(2))


> Moreover, datesync is not a good var name because it's not a date.

I learned a lot of readability versus comments.

I have a tendency of over commenting because I often had to deal with
code I did not write and thus had to understand. To that end it often
ended up with renaming functions to better semanticc.

I have always strived to have semantic names for my variables and
functions - so your suggestion is quite in my spirit :)

Also the format - I will add your proposition - it's a good one.

fhdk commented:
ons, 11 01 2017 kl. 11:45 -0800, skrev Hugo Posnic:
> If we try to follow the versioning conventions, I think that changes
> that has been realized require at least a 2.1 version or make a bump
> to 3.0 and not 2.0.1.

I had a look at the link and I must say - the versioning has never been
my strong side. Mostly just incrementing the build version. Hence 2.0.1
:)

I think - with the semantic in mind - the compared changes are big
enough for a version 3.0 - if one of the original developers looked at
code now, they would propably say its not their code.

Do you agree? @kirek

Huluti commented:
Very good changes, but like I've said don't do not worry about the version number. Come back as before. Kirek or I will take care of it. A version jump is done in a separated commit  by the maintainers of the directory.

Huluti commented:
I've just merged the UI PR. Think to [rebase this PR.](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) ;).

Kirek commented:
I think this is good enough to merge if there are some bugs we can fix them later. Good job everyone!
